### PR TITLE
Add MKLDNNDeviceContext

### PR DESCRIPTION
**Pull request by tensor-tang**
Add mkldnn device context and mkldnn_helper

### CODE REVIEW

**Comment by tensor-tang** [comment]
Still have deps issue:

>[11:23:21]W:	 [Step 1/1] In file included from /paddle/paddle/platform/device_context.h:25:0,
[11:23:21]W:	 [Step 1/1]                  from /paddle/paddle/framework/tensor.h:26,
[11:23:21]W:	 [Step 1/1]                  from /paddle/paddle/framework/lod_tensor.h:26,
[11:23:21]W:	 [Step 1/1]                  from /paddle/paddle/framework/operator.h:27,
[11:23:21]W:	 [Step 1/1]                  from /paddle/paddle/framework/grad_op_desc_maker.h:20,
[11:23:21]W:	 [Step 1/1]                  from /paddle/paddle/framework/shape_inference.cc:15:
[11:23:21]W:	 [Step 1/1] /paddle/paddle/platform/mkldnn_helper.h:17:22: fatal error: mkldnn.hpp: No such file or directory
[11:23:21]W:	 [Step 1/1] compilation terminated.

**Comment by tensor-tang** [comment]
Found that the above issue is caused by `cc_library(shape_inference SRCS shape_inference.cc DEPS ddim attribute)` missing depends on `device_context`

**Comment by tensor-tang** [comment]
The last lib issue is fixed by dummy target.

**Comment by QiJune** [comment]
What's the file /paddle/operators/tensor.save used for?

File: paddle/platform/device_context.h [hunk 1]
```diff
@@ -117,6 +121,54 @@ class CUDNNDeviceContext : public CUDADeviceContext {
...
+  template <typename T>
+  const T GetElement(const std::string& op_key) const;
+
+  /* \brief  Get element pool: memory, primitive or primitive desc pool */
+  template <typename T>
+  const std::unordered_map<const std::string, const T, std::hash<std::string>>&
+  GetElementPool() const;
+
+  /* \brief  Get the active engine */
+  const MKLDNNEnginePtr GetEngine() const { return engine_; }
```
Comment (QiJune) [inline paddle/platform/device_context.h]: GetEngine --> engine

File: paddle/platform/mkldnn_helper.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+namespace paddle {
+namespace platform {
+
+using MKLDNNStream = mkldnn::stream;
+using MKLDNNEngine = mkldnn::engine;
+using MKLDNNMemory = mkldnn::memory;
+using MKLDNNPrimitive = mkldnn::primitive;
+using MKLDNNPrimitiveDesc = mkldnn::handle<mkldnn_primitive_desc_t>;
+
+typedef std::shared_ptr<MKLDNNEngine> MKLDNNEnginePtr;
```
Comment (QiJune) [inline paddle/platform/mkldnn_helper.h]: I am not sure if std::shared_ptr is necessary, can we use unique_ptr here?

**Comment by tensor-tang** [comment]
> What's the file /paddle/operators/tensor.save used for?

Weird, I did not add this file. 

Will remove it, thanks for catching this.

File: paddle/platform/device_context.h (hunk 1 above)
Comment (tensor-tang) [reply paddle/platform/device_context.h]: OK, thx~

File: paddle/platform/mkldnn_helper.h (hunk 2 above)
Comment (tensor-tang) [reply paddle/platform/mkldnn_helper.h]: OK, I think it's fine.

**Comment by tensor-tang** [comment]
> 294 - test_dynrnn_gradient_check (Failed)

I will merge latest.

**Comment by dzhwinter** [comment]
> test_dynrnn_gradient_check

This test is added by @reyoung  recently, which is fragile in CI build, we can add a bigger threshold in current stage and fix later.

**Comment by tensor-tang** [comment]
Thanks for reminder @dzhwinter

**Comment by tensor-tang** [comment]
Failed with the same test `test_dynrnn_gradient_check`

I will rerun PR_CI.

**Pull request merged**